File: java/1544. Make The String Great.java
class Solution {
    public String makeGood(String s) {
        Deque<Character> stack = new ArrayDeque<>();
        for (int i = 0; i < s.length(); i++) {
            if (stack.isEmpty()) {
                stack.addLast(s.charAt(i));
                continue;
            }
            if (checkSame(stack.peekLast(), s.charAt(i))) {
                stack.pollLast();
            } else {
                stack.addLast(s.charAt(i));
            }
        }
        StringBuilder sb = new StringBuilder();
        while (!stack.isEmpty()) {
            sb.append(stack.pollFirst());
        }
        return sb.toString();
    }
    private boolean checkSame(char a, char b) {
        if (a <= 'Z' && a >= 'A') {
            return a-'A' == b-'a';
        }
        if (a <= 'z' && a >= 'a') {
            return a-'a' == b-'A';
        }
        return false;
    }
}

File: go/1544. Make The String Great.go
func makeGood(s string) string {
	var stack []byte
	for i := 0; i < len(s); i++ {
		if len(stack) == 0 {
			stack = append(stack, s[i])
			continue
		}
		if checkSame(stack[len(stack)-1], s[i]) {
			stack = stack[:len(stack)-1]
		} else {
			stack = append(stack, s[i])
		}
	}
	return string(stack)
}

func checkSame(a, b byte) bool {
	if a >= 'a' && a <= 'z' {
		return a-'a' == b-'A'
	}
	if a >= 'A' && a <= 'Z' {
		return a-'A' == b-'a'
	}
	return false
}